server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java

```diff
@@ -36,417 +40,479 @@
+                        allocation.changes()
+                    );
+                }
             }
+        }
 
-            final var routingNode = routingNodes.node(shardRouting.currentNodeId());
-            final var canRemainDecision = allocation.deciders().canRemain(shardRouting, routingNode, allocation);
-            if (canRemainDecision.type() != Decision.Type.NO) {
-                // it's desired elsewhere but technically it can remain on its current node. Defer its movement until later on to give
-                // priority to shards that _must_ move.
-                continue;
+        private void balance() {
+            if (allocation.deciders().canRebalance(allocation).type() != Decision.Type.YES) {
+                return;
             }
 
-            final var moveTarget = findRelocationTarget(shardRouting, assignment.nodeIds());
-            if (moveTarget != null) {
-                routingNodes.relocateOrReinitializeShard(
-                    shardRouting,
-                    moveTarget.getId(),
-                    allocation.clusterInfo().getShardSize(shardRouting, ShardRouting.UNAVAILABLE_EXPECTED_SHARD_SIZE),
-                    allocation.changes()
-                );
-            }
-        }
-    }
+            long allAllocations = 0;
+            long undesiredAllocations = 0;
+            // Iterate over the started shards interleaving between nodes, and try to move any which are on undesired nodes. In the presence
+            // of throttling shard movements, the goal of this iteration order is to achieve a fairer movement of shards from the nodes that
+            // are offloading the shards.
+            for (final var iterator = routingNodes.nodeInterleavedShardIterator(); iterator.hasNext();) {
+                final var shardRouting = iterator.next();
+
+                if (shardRouting.started() == false) {
+                    // can only rebalance started shards
+                    continue;
+                }
 
-    private void balance() {
-        if (allocation.deciders().canRebalance(allocation).type() != Decision.Type.YES) {
-            return;
-        }
+                final var assignment = desiredBalance.getAssignment(shardRouting.shardId());
+                if (assignment == null) {
+                    // balance is not computed
+                    continue;
+                }
 
-        // Iterate over the started shards interleaving between nodes, and try to move any which are on undesired nodes. In the presence of
-        // throttling shard movements, the goal of this iteration order is to achieve a fairer movement of shards from the nodes that are
-        // offloading the shards.
-        for (final var iterator = routingNodes.nodeInterleavedShardIterator(); iterator.hasNext();) {
-            final var shardRouting = iterator.next();
+                allAllocations++;
```

Comment:
This only counts STARTED shards with a desired location, which I think might be confusing. Let's count every iteration in the loop. (If you want, we can reasonably also count & report the number of non-STARTED shards and shards without a desired location).